src/fu-engine.c

```diff
@@ -221,20 +221,70 @@ fu_engine_item_free (FuDeviceItem *item)
 	g_free (item);
 }
 
+static gsize
+fu_engine_get_prefix_size (const gchar *hash, const gchar *hash_partial)
+{
+	gsize i;
+	/* returns the number of chars that match from the start, e.g.
+	 * deadbeef,deadf00d=>4 and deadbeef,f00dbeef=>0 */
+	for (i = 0; hash[i] != '\0' && hash_partial[i] != '\0'; i++) {
+		if (hash[i] != hash_partial[i])
+			return i;
+	}
+	return i;
+}
+
 static FuDeviceItem *
 fu_engine_get_item_by_id (FuEngine *self, const gchar *device_id, GError **error)
 {
-	for (guint i = 0; i < self->devices->len; i++) {
+	FuDeviceItem *item_best = NULL;
+	gboolean multiple_matches = FALSE;
+	gsize device_id_len = strlen (device_id);
+	gsize sz_best = 0;
+
+	/* support abbreviated hashes */
+	for (gsize i = 0; i < self->devices->len; i++) {
 		FuDeviceItem *item = g_ptr_array_index (self->devices, i);
-		if (g_strcmp0 (fu_device_get_id (item->device), device_id) == 0)
-			return item;
-		if (g_strcmp0 (fu_device_get_equivalent_id (item->device), device_id) == 0)
-			return item;
+		const gchar *ids[] = {
+			fu_device_get_id (item->device),
+			fu_device_get_equivalent_id (item->device),
+			NULL };
+		for (guint j = 0; ids[j] != NULL; j++) {
+			gsize sz = fu_engine_get_prefix_size (ids[j], device_id);
+			if (sz < device_id_len)
+				continue;
+			if (sz > sz_best) {
+				sz_best = sz;
+				item_best = item;
+				multiple_matches = FALSE;
+				continue;
+			}
+			if (sz == sz_best) {
+				multiple_matches = TRUE;
+				continue;
+			}
+		}
+	}
+
+	/* multiple things matched */
+	if (item_best != NULL && multiple_matches) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "device ID %s was not unique",
+			     device_id);
+		return FALSE;
 	}
+
+	/* something good found */
+	if (item_best != NULL)
+		return item_best;
+
+	/* nothing at all matched */
 	g_set_error (error,
 		     FWUPD_ERROR,
-		     FWUPD_ERROR_INVALID_FILE,
-		     "Device %s was not found",
+		     FWUPD_ERROR_NOT_FOUND,
+		     "device ID %s was not found",
 		     device_id);
 	return NULL;
```

Comment:
If an item is found `item_best `means something, if not it's `NULL`.
How about this instead:
```
if (item_best == NULL)
        g_set_error (...)

return item_best;
```